### DISCUSSION TITLE
Fire PRE_OPERATIONS_EXEC per batch, allowing plugins to hook into it?

### DISCUSSION THREAD

**Discussion by dzuelke** [composer/composer#10365] [General]
With `plugin-modifies-downloads`, Composer 2 runs the install in batches: first, the plugins, then, all other packages.

Wouldn't this, in theory, allow dispatching `PRE_OPERATIONS_EXEC` multiple times, once for each batch?

Right now, a plugin cannot hook into this event, because it's not installed yet by the time this event fires.

However, on a second `composer install` run, the plugin is there, and now hooks into the event correctly, which IMO can make for very inconsistent or unexpected behavior, although I'd assume (or hope) nobody would really ship such a plugin.

Should this work for plugins with `plugin-modifies-downloads` (which I guess would then turn into `plugin-modifies-behavior` really)?

If not, would it maybe make sense to never register this event for non-global plugins?

**Comment by Seldaek** [comment]
> Should this work for plugins with plugin-modifies-downloads (which I guess would then turn into plugin-modifies-behavior really)?
 
Not really.. I mean it's not expected to, and it'd be a mess to implement.

> If not, would it maybe make sense to never register this event for non-global plugins?

That would be sensible, but also quite hard to implement without adding some weird plugin-specific code to EventDispatcher.

So all in all I wonder if we shouldn't just leave it as is, and perhaps make sure this is documented better in https://getcomposer.org/doc/articles/scripts.md#installer-events to avoid anyone using that event without recommending global plugin installation.

**Reply by naderman** [reply]
No strong opinion, but think I agree on just documenting this better, if there isn't a good important use case to change this?

**Reply by Seldaek** [reply]
Fixed (hopefully clearly enough) by acea849c513b5195c25587982ed9f3bd497d6473
